src/runtime_src/core/pcie/tools/cloud-daemon/common.cpp

```diff
@@ -144,91 +161,88 @@ int waitForMsg(pcieFunc& dev, int localfd, int remotefd, long interval)
     if (ret == 0)
         return -EAGAIN; // time'd tout
 
-    if (FD_ISSET(localfd, &fds)) {
-        retfd = localfd;
-        dev.log(LOG_INFO, "msg arrived on mailbox fd %d", retfd);
-    } else {
-        retfd = remotefd;
-        dev.log(LOG_INFO, "msg arrived on remote fd %d", retfd);
+    //it is possible both FDs have data ready concurrently.
+    if (localfd > 0 && FD_ISSET(localfd, &fds)) {
+        retfd[0] = localfd;
+        dev.log(LOG_INFO, "msg arrived on mailbox fd %d", retfd[0]);
+    }
+    if (remotefd > 0 && FD_ISSET(remotefd, &fds)) {
+        retfd[1] = remotefd;
+        dev.log(LOG_INFO, "msg arrived on remote fd %d", retfd[1]);
     }
-    return retfd;
+    return 0;
 }
 
 /*
- * Fetch sw channel msg from local mailbox fd, process it by passing it through
- * to socket fd or by the callback.
+ * Fetch sw channel msg from local mailbox fd
  */
-int processLocalMsg(pcieFunc& dev, int localfd, int remotefd, msgHandler cb)
+std::shared_ptr<sw_msg> getLocalMsg(pcieFunc& dev, int localfd)
 {
     size_t msgsz = getMailboxMsgSize(dev, localfd);
     if (msgsz == 0)
-        return -EINVAL;
+        return nullptr;
 
     std::shared_ptr<sw_msg> swmsg = std::make_shared<sw_msg>(msgsz);
     if (swmsg == nullptr)
-        return -ENOMEM;
+        return nullptr;
 
-    if (!readMsg(dev, localfd, swmsg.get()))
-        return -EINVAL;
+    if (readMsg(dev, localfd, swmsg.get()) != 0)
+        return nullptr;
 
-    int pass;
-    std::shared_ptr<sw_msg> swmsgProcessed;
-    if (!cb) {
-        // Continue passing received msg to local mailbox.
-        swmsgProcessed = swmsg;
-        pass = FOR_REMOTE;
-    } else {
-        pass = (*cb)(dev, swmsg, swmsgProcessed);
-    }
-
-    bool sent;
-    if (pass == FOR_LOCAL)
-        sent = sendMsg(dev, localfd, swmsgProcessed.get());
-    else if (pass == FOR_REMOTE)
-        sent = sendMsg(dev, remotefd, swmsgProcessed.get());
-    else // Error occured
-        return pass;
-
-    return sent ? 0 : -EINVAL;
+    return swmsg;
```

Comment:
It looks like you want a `std::optional` here (or `boost::optional` if you are forced to use old C++).

Reply:
is optional c++17? 
we need to make sure it work on c++14.

Reply:
So `boost::optional` then...

Reply:
what is the benefit to use optional here?